Lib/test/_test_multiprocessing.py

```diff
@@ -3610,6 +3616,246 @@ def test_copy(self):
         self.assertAlmostEqual(bar.y, 5.0)
         self.assertEqual(bar.z, 2 ** 33)
 
+
+class _TestSharedMemory(BaseTestCase):
+
+    ALLOWED_TYPES = ('processes',)
+
+    def setUp(self):
+        if not HAS_SHMEM:
+            self.skipTest("requires multiprocessing.shared_memory")
+
+    @staticmethod
+    def _attach_existing_shmem_then_write(shmem_name, binary_data):
+        local_sms = shared_memory.SharedMemory(shmem_name)
+        local_sms.buf[:len(binary_data)] = binary_data
+        local_sms.close()
+
+    def test_shared_memory_basics(self):
+        sms = shared_memory.SharedMemory(
+            'test01_tsmb',
+            flags=shared_memory.O_CREX,
+            size=512
+        )
+        try:
+            # Verify attributes are readable.
+            self.assertEqual(sms.name, 'test01_tsmb')
+            self.assertGreaterEqual(sms.size, 512)
+            self.assertGreaterEqual(len(sms.buf), sms.size)
+            self.assertEqual(sms.mode, 0o600)
+
+            # Modify contents of shared memory segment through memoryview.
+            sms.buf[0] = 42
+            self.assertEqual(sms.buf[0], 42)
+
+            # Attach to existing shared memory segment.
+            also_sms = shared_memory.SharedMemory('test01_tsmb')
+            self.assertEqual(also_sms.buf[0], 42)
+            also_sms.close()
+
+            if isinstance(sms, shared_memory.PosixSharedMemory):
+                # Posix Shared Memory can only be unlinked once.  Here we
+                # test an implementation detail that is not observed across
+                # all supported platforms (since WindowsNamedSharedMemory
+                # manages unlinking on its own and unlink() does nothing).
+                # True release of shared memory segment does not necessarily
+                # happen until process exits, depending on the OS platform.
+                with self.assertRaises(shared_memory.ExistentialError):
+                    sms_uno = shared_memory.SharedMemory(
+                        'test01_dblunlink',
+                        flags=shared_memory.O_CREX,
+                        size=5000
+                    )
+
+                    try:
+                        self.assertGreaterEqual(sms_uno.size, 5000)
+
+                        sms_duo = shared_memory.SharedMemory('test01_dblunlink')
+                        sms_duo.unlink()  # First shm_unlink() call.
+                        sms_duo.close()
+                        sms_uno.close()
+
+                    finally:
+                        sms_uno.unlink()  # A second shm_unlink() call is bad.
```

Comment:
also here (could use `addCleanup`)

Reply:
Actually, this one needs to stay as it is, because this second call to `unlink()` is needed to trigger the exception that we are testing for.